test-harness/src/checkpoint.rs

```diff
@@ -0,0 +1,104 @@
+use std::fmt::Debug;
+use std::fmt::Display;
+use std::ops::Range;
+use std::ops::RangeFrom;
+use std::ops::RangeFull;
+use std::ops::RangeInclusive;
+use std::ops::RangeTo;
+use std::ops::RangeToInclusive;
+use std::slice::SliceIndex;
+
+use crate::Event;
+
+/// A checkpoint in a `ghcid-ng` run.
+///
+/// [`crate::GhcidNg`] provides methods for asserting that events are logged, or waiting for
+/// events to be logged in the future.
+///
+/// To avoid searching thousands of log events for each assertion, and to provide greater
+/// granularity for assertions, you can additionally assert that events are logged between
+/// particular checkpoints.
+///
+/// Checkpoints can be constructed with [`crate::GhcidNg::first_checkpoint`],
+/// [`crate::GhcidNg::current_checkpoint`], and [`crate::GhcidNg::checkpoint`].
+#[derive(Debug, Clone, Copy)]
+pub struct Checkpoint(pub(crate) usize);
```

Comment:
I personally feel like this `Checkpoint` "newtype" doesn't buy you much over just using `usize` directly and the `Checkpoint` type adds a lot of complexity here.  Most of the `CheckpointIndex` `impl`s don't really do anything other than `coerce` `Checkpoint` to `usize`

Reply:
It's a bounds-checked index, so it's guaranteed to be safe to use. If we used a `usize` the user could pass in garbage and cause the test to panic. Probably not that bad, but we _do_ get a safety guarantee out of this.

Reply:
Oh, actually, the thing it does give us is a `SliceIndex` implementation with slice output for single checkpoints. This lets us treat single checkpoints like `1` and ranges like `1..3` or `2..` the same. The standard `SliceIndex` implementations instead return a `T` for a `usize` and a `[T]` for ranges of `usize`.

So we'd need to keep the `CheckpointIndex` trait around, even if it was just implemented for `usize` and ranges of `usize`.